metadata-ingestion/src/datahub/ingestion/source/sql/clickhouse.py

```diff
@@ -118,30 +124,92 @@ def __post_init__(self):
 
 
 class ClickHouseConfig(
-    BasicSQLAlchemyConfig, BaseTimeWindowConfig, DatasetLineageProviderConfigBase
+    TwoTierSQLAlchemyConfig, BaseTimeWindowConfig, DatasetLineageProviderConfigBase
 ):
     # defaults
     host_port = Field(default="localhost:8123", description="ClickHouse host URL.")
     scheme = Field(default="clickhouse", description="", hidden_from_docs=True)
     password: pydantic.SecretStr = Field(
         default=pydantic.SecretStr(""), description="password"
     )
-
     secure: Optional[bool] = Field(default=None, description="")
     protocol: Optional[str] = Field(default=None, description="")
+    _deprecate_secure = pydantic_field_deprecated("secure")
+    _deprecate_protocol = pydantic_field_deprecated("protocol")
 
+    uri_opts: Dict[str, str] = Field(
+        default={},
+        description="The part of the URI and it's used to provide additional configuration options or parameters for the database connection.",
+    )
     include_table_lineage: Optional[bool] = Field(
         default=True, description="Whether table lineage should be ingested."
     )
     include_materialized_views: Optional[bool] = Field(default=True, description="")
 
-    def get_sql_alchemy_url(self, database=None):
-        uri_opts = None
-        if self.scheme == "clickhouse+native" and self.secure:
-            uri_opts = {"secure": "true"}
-        elif self.scheme != "clickhouse+native" and self.protocol:
-            uri_opts = {"protocol": self.protocol}
-        return super().get_sql_alchemy_url(uri_opts=uri_opts)
+    def get_sql_alchemy_url(self, current_db=None):
+        url = make_url(
+            super().get_sql_alchemy_url(uri_opts=self.uri_opts, current_db=current_db)
+        )
+        if url.drivername == "clickhouse+native" and url.query.get("protocol"):
+            logger.debug(f"driver = {url.drivername}, query = {url.query}")
+            raise Exception(
+                "You cannot use a schema clickhouse+native and clickhouse+http at the same time"
+            )
+
+        self.scheme = url.drivername
+        self.username = url.username
+        self.password = (
+            pydantic.SecretStr(str(url.password))
+            if url.password
+            else pydantic.SecretStr("")
+        )
+        if url.host and url.port:
+            self.host_port = url.host + ":" + str(url.port)
+        elif url.host:
+            self.host_port = url.host
+        # untill released https://github.com/python/mypy/pull/15174
+        self.uri_opts = {str(k): str(v) for (k, v) in url.query.items()}
+
+        url = make_url(
+            make_sqlalchemy_uri(
+                self.scheme,
+                self.username,
+                self.password.get_secret_value() if self.password else None,
+                self.host_port,
+                current_db if current_db else self.database,
+                uri_opts=self.uri_opts,
+            )
+        )
```

Comment:
From what I understand this special handling is required for the case when `sqlalchemy_uri` is set and `database` is not set - i.e. the case when multiple databases are to be ingested by specifying sqlalchemy_uri.

This combination is also not supported for other "two tier sqlalchemy sources" (mysql, hive, etc). An ideal solution would take care of above condition for all sources, however I am okay to handle this only for clickhouse here. Not very fond of the way code had to be written to keep it compatible with sqlalchemy 1.3.

Can we move this code inside if condition -> `if self.sqlalchemy_uri and not self.database: ` and add a comment about this ? It'll help in future to understand why this had to be done.